### PR TITLE
Add VK_KHR_shader_atomic_int64

### PR DESCRIPTION
**Pull request by sfricke-samsung**
First part of #2629

I have the rest of the atomics ready to go, but wanted to test the waters with the first (simplest) check first before adding the rest

### CODE REVIEW

File: layers/shader_validation.cpp [hunk 1]
```diff
@@ -1455,6 +1455,33 @@ bool CoreChecks::ValidateShaderResolveQCOM(SHADER_MODULE_STATE const *src, VkPip
...
+            // Validate 64-bit atomics
+            if (((atomic.storage_class == spv::StorageClassStorageBuffer) || (atomic.storage_class == spv::StorageClassUniform)) &&
+                (enabled_features.core12.shaderBufferInt64Atomics == VK_FALSE)) {
+                skip |= LogError(
+                    device, "UNASSIGNED-CoreValidation-Shader-shaderBufferInt64Atomics",
+                    "%s: Can't use 64-bit int atomics operations with %s storage class without shaderBufferInt64Atomics enabled.",
+                    report_data->FormatHandle(src->vk_shader_module()).c_str(), StorageClassName(atomic.storage_class));
+            } else if ((atomic.storage_class == spv::StorageClassWorkgroup) &&
+                       (enabled_features.core12.shaderSharedInt64Atomics == VK_FALSE)) {
+                skip |= LogError(device, "UNASSIGNED-CoreValidation-Shader-shaderSharedInt64Atomics",
```
Comment (ncesario-lunarg) [inline layers/shader_validation.cpp:1474]: Is this defined somewhere? I would vote for using a constant here (ditto for above).

File: tests/vklayertests_pipeline_shader.cpp [hunk 2]
```diff
@@ -11543,3 +11543,114 @@ TEST_F(VkLayerTest, PipelineSubgroupSizeControl) {
...
+        m_instance_extension_names.push_back(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+    } else {
+        printf("%s Did not find required instance extension %s; skipped.\n", kSkipPrefix,
+               VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+        return;
+    }
+
+    // Create device without VK_KHR_shader_atomic_int64 extension or features enabled
+    ASSERT_NO_FATAL_FAILURE(InitFramework(m_errorMonitor));
+    if (version < VK_API_VERSION_1_1) {
```
Comment (ncesario-lunarg) [inline tests/vklayertests_pipeline_shader.cpp:11561]: Does this need a device >= 1.1, or just instance >= 1.1?

**Review by ncesario-lunarg — CHANGES_REQUESTED** [review]
Looks like an "unassigned VUID" needs to be added to `core_validation_error_enums.h`, but other than that, LGTM!

File: layers/shader_module.cpp [hunk 3]
```diff
@@ -434,6 +438,33 @@ void SHADER_MODULE_STATE::BuildDefIndex() {
             } break;
 
             default:
+                // Because atomics operands all need to be defined can do in initial pass
```
[the remark this answers is not in the document]
Comment (sfricke-samsung) [reply layers/shader_module.cpp:441]: This is a useless comment, SPIR-V is SSA (and the forward declaration stuff is not related) so this is an obvious assumption

File: tests/vklayertests_pipeline_shader.cpp (hunk 2 above)
Comment (sfricke-samsung) [reply tests/vklayertests_pipeline_shader.cpp:11561]: I need SPIR-V 1.3, so `device`, will fix

File: layers/shader_module.cpp [hunk 4]
```diff
@@ -434,6 +438,33 @@ void SHADER_MODULE_STATE::BuildDefIndex() {
...
+                    // spirv-val should catch if not pointer
+                    assert(pointer.opcode() == spv::OpTypePointer);
+                    atomic.storage_class = pointer.word(2);
+
+                    auto data_type = get_def(pointer.word(3));
+                    atomic.type = data_type.opcode();
+
+                    // TODO - Should have a proper GetBitWidth like spirv-val does
+                    assert(data_type.opcode() == spv::OpTypeFloat || data_type.opcode() == spv::OpTypeInt);
+                    atomic.bit_width = data_type.word(2);
```
[the remark this answers is not in the document]
Comment (sfricke-samsung) [reply layers/shader_module.cpp:464]: There are many other places I would use `GetBitWidth` as well

I was planning on making a follow up PR with more "utils" like this and populate other places

**Comment by sfricke-samsung** [comment]
unable to use git it seems today, sorry for launching like 5 CI builds there

File: layers/shader_module.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncesario-lunarg) [reply layers/shader_module.cpp:463]: :+1: if you do before I get a chance to finish my other change, I'll use them. Though, adding all of these "utility functions" makes me want to investigate using something like spirv-cross more...

**Review by ncesario-lunarg — APPROVED** [review]

**Pull request merged**